junit-plugin/src/main/scala/org/scalajs/junit/plugin/ScalaJSJUnitPlugin.scala

```diff
@@ -0,0 +1,447 @@
+package org.scalajs.junit.plugin
+
+import scala.language.reflectiveCalls
+
+import org.scalajs.junit._
+
+import scala.reflect.internal.Flags
+import scala.tools.nsc._
+import scala.tools.nsc.plugins.{
+Plugin => NscPlugin, PluginComponent => NscPluginComponent
+}
+
+class ScalaJSJUnitPlugin(val global: Global) extends NscPlugin {
+  selfPlugin =>
+
+  val name: String = "Scala.js JUnit plugin"
+
+  val components: List[NscPluginComponent] =
+    List(ScalaJSJUnitPluginComponent)
+
+  val description: String = "Adapts JUnit test classes to Scala.js"
+
+  object ScalaJSJUnitPluginComponent extends plugins.PluginComponent
+      with transform.Transform {
+
+    protected def newTransformer(
+        unit: global.CompilationUnit): global.Transformer = {
+      new ScalaJSJUnitPluginTransformer
+    }
+
+    val global: Global = selfPlugin.global
+    val phaseName: String = "junit-inject"
+    val runsAfter: List[String] = List("mixin")
+    override val runsBefore: List[String] = List("jscode")
+
+    class ScalaJSJUnitPluginTransformer extends global.Transformer {
+
+      import global._
+
+      type ScalaJSPlugin = NscPlugin {
+        def PrepInteropComponent: AnyRef {
+          def genExportMember(sym: Symbol): List[Tree]
+          def registerModuleExports(sym: Symbol): Unit
+          def genNamedExport(defSym: Symbol, jsName: String, pos: Position): Unit
+        }
+      }
```

Comment:
Adding the annotations at this point is too late. Exports of classes and modules have already been recorded by prepJSInterop and stored in internal `Set[Symbol]`s through `registerModuleExports`.

Having this plugin run before `mixin` would make it much much harder to guarantee that we emit the right metadata. After all, we're trying to mimic what Java reflection sees. The closer to the back-end, the better. Before erasure, it would be hopeless to try.